// @tanstack/start-storage-context/dist/esm/async-local-storage.js
import * as asyncHooks from "node:async_hooks";
const AsyncLocalStorage = asyncHooks.AsyncLocalStorage || class {
  constructor() {
    this.store = undefined;
  }
  getStore() {
    return this.store;
  }
  run(store, fn) {
    const previous = this.store;
    this.store = store;
    try {
      return fn();
    } finally {
      this.store = previous;
    }
  }
};
//#region src/async-local-storage.ts
var GLOBAL_STORAGE_KEY = Symbol.for("tanstack-start:start-storage-context");
var globalObj = globalThis;
if (!globalObj[GLOBAL_STORAGE_KEY]) globalObj[GLOBAL_STORAGE_KEY] = new AsyncLocalStorage();
var startStorage = globalObj[GLOBAL_STORAGE_KEY];
async function runWithStartContext(context, fn) {
	return startStorage.run(context, fn);
}
function getStartContext(opts) {
	const context = startStorage.getStore();
	if (!context && opts?.throwIfNotFound !== false) throw new Error(`No Start context found in AsyncLocalStorage. Make sure you are using the function within the server runtime.`);
	return context;
}
//#endregion
export { getStartContext, runWithStartContext };
